.github/workflows/haskell.yml

```diff
@@ -46,138 +43,171 @@ jobs:
+
+    - name: "LINUX: Install secp256k1"
+      if: runner.os != 'Windows'
+      shell: bash
+      env:
+        CI_SECP_FLAGS: "--prefix=/usr/local"
+        CI_SECP_INSTALL_CMD: sudo
+      run: bash .github/workflows/build-secp256k1.bash
+
+    # TODO: this really should come from a pre-built location
+    - name: "WIN: Install secp256k1"
+      if: runner.os == 'Windows'
+      # Same env as tmate action
+      env:
+        MSYS2_PATH_TYPE: inherit
+        MSYSTEM: MINGW64
+        CHERE_INVOKING: 1
+        # install secp into /mingw64 prefix, which is where pkg-config will look
+        # by default.
+        CI_SECP_FLAGS: "--prefix=/mingw64"
+      run: C:\\msys64\\usr\\bin\\bash.exe .github/workflows/build-secp256k1.bash
 
     - name: Cabal update
       run: cabal update
 
-    - name: Cabal Configure
-      run: cabal configure --enable-tests --enable-benchmarks --write-ghc-environment-files=always
+    - name: Configure build
+      shell: bash
+      run: |
+        if [ "${{github.event.inputs.tests}}" == "all" ]; then
+          echo "Reconfigure cabal projects to run tests for all dependencies"
+          sed -i 's|tests: False|tests: True|g' cabal.project
+        fi
+
+        cp ".github/workflows/cabal.project.local.ci.$(uname -s)" cabal.project.local
+
+        echo "# cabal.project.local"
+        cat cabal.project.local
 
     - name: Record dependencies
+      id: record-deps
       run: |
+        cabal build all --dry-run
         cat dist-newstyle/cache/plan.json | jq -r '."install-plan"[].id' | sort | uniq > dependencies.txt
 
-    - name: Set cache version
-      run: echo "CACHE_VERSION=9w76Z3Q" >> $GITHUB_ENV
+    - name: "OUTPUT Record weeknum"
+      shell: bash
+      run: echo "weeknum=$(/usr/bin/date -u "+%W")" >> $GITHUB_OUTPUT
 
-    - uses: actions/cache@v2
-      name: Cache cabal store
+    - name: Cache Cabal store
+      uses: actions/cache@v2
       with:
-        path: ${{ steps.setup-haskell.outputs.cabal-store }}
-        key: cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('dependencies.txt') }}
+        path: ${{ runner.os == 'Windows' && steps.win-setup-haskell.outputs.cabal-store || steps.lin-setup-haskell.outputs.cabal-store }}
+        key: cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('dependencies.txt') }}-${{ hashFiles('date.txt') }}
         restore-keys: |
           cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('dependencies.txt') }}
-          cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-
+          cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}
+
+    - uses: actions/cache@v2
+      name: "Cache `dist-newstyle`"
+      with:
+        path: |
+          dist-newstyle
+          !dist-newstyle/**/.git
+        key: cache-dist-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-${{ steps.record-deps.outputs.weeknum }}
```

Comment:
@angerman

Reply:
I wish I knew, this was pre-existing code. Ultimately we probably want to cache hash(env) to ensure that changes to the environment don't give us a stale cache.

Reply:
> Ultimately we probably want to cache hash(env) to ensure that changes to the environment don't give us a stale cache.

Any change to the environment suppose to be accompanied by a change to `env.CACHE_VERSION`. Weeknum isn't gonna solve a change to environment at the worst case for a whole week. So, I don't buy that argument.

That suffix means that we store a totally new cache every week. But we can never use older cache, because the ordering of the key will always choose the newer week.

So, this seems to me like a totally useless logic. I suggest we remove the `weeknum` step and the suffix `-${{ steps.record-deps.outputs.weeknum }}`. Especially since no-one knows what is it for. 

I personally can't see how it is useful.

Reply:
I think I understand now.  This is for caching `dist-newstyle` and the week cycling means we have a different cache per week.  I'm skeptical it works though.

I think we can safely drop the entire step since we're moving to CHaP so there should be much less stuff in there.